File: robot-seed/src/test/clojure/liu/mars/market/listener_test.clj
(ns liu.mars.market.listener-test
  (:require [clojure.test :refer :all]
            [jaskell.handle :refer [supplier function]])
  (:import (akka.actor ActorSystem ActorPaths)
           (akka.testkit.javadsl TestKit)
           (liu.mars.market.dash Depth)
           (akka.cluster.client ClusterClientReceptionist ClusterClientSettings ClusterClient ClusterClient$Send ClusterClient$SendToAll ClusterClient$Publish)
           (akka.cluster.pubsub DistributedPubSub DistributedPubSubMediator$Publish DistributedPubSubMediator DistributedPubSubMediator$Subscribe DistributedPubSubMediator$SubscribeAck DistributedPubSubMediator$Unsubscribe)
           (akka.cluster Cluster)
           (java.util.function Function)
           (liu.mars ClojureActor)))

(defn initer [actor]
  (send (.getState actor)
        #(assoc % :feedback (-> self
                                (.path)
                                (.name)))))

(deftest direct-feedback-test
  "startup a inner actor and link to cluster, publish message and get feedback"
  (let [channel "btcusdt.depth.step0"
        system (ActorSystem/create "market")
        test-kit (TestKit. system)
        self (.getRef test-kit)
        empty-depth (doto (Depth.)
                      (.setChannel "btcusdt.depth.step0")
                      (.setVersion 0)
                      (.setAsk [])
                      (.setBid []))
        await #(.awaitCond test-kit (supplier [_] (.msgAvailable test-kit)))
        cluster (Cluster/get system)
        mediator (-> system
                     (DistributedPubSub/get)
                     (.mediator))]
    (try
      (-> (DistributedPubSubMediator$Subscribe. channel self)
          (#(.tell mediator % self)))
      (await)
      (.expectMsgPF test-kit "subscribe ack"
                    (function [msg]
                      (is (instance? DistributedPubSubMediator$SubscribeAck msg))))
      (-> mediator
          (.tell (DistributedPubSubMediator$Publish. channel, empty-depth) self))
      (await)
      (.expectMsgPF test-kit "expect empty depth"
                    (function [msg]
                      (is (and (instance? Depth msg)
                               (= empty-depth msg)))))
      (-> (DistributedPubSubMediator$Unsubscribe. channel self)
          (#(.tell mediator % self)))
      (await)
      (finally
        (do
          (.leave cluster (.selfAddress cluster))
          (TestKit/shutdownActorSystem system)
          (Thread/sleep 1000))))))

(defmulti service (fn [_ msg] (class msg)))
(defmethod service Depth [this msg]
  (-> this
      (.context)
      (.system)
      (.actorSelection (str "/system/" (:feedback @(.getState this))))
      (.tell msg (.self this))))

(deftest client-feedback-test
  "startup a inner actor and link to cluster, post messages from client and get feedback"
  (let [channel "btcusdt.depth.step0"
        system (ActorSystem/create "market")
        test-kit (TestKit. system)
        self (.getRef test-kit)
        port (-> system
                 (.provider)
                 (.getDefaultAddress)
                 (.port)
                 (.get))
        empty-depth (doto (Depth.)
                      (.setChannel "btcusdt.depth.step0")
                      (.setVersion 0)
                      (.setAsk [])
                      (.setBid []))
        await #(.awaitCond test-kit (supplier [_] (.msgAvailable test-kit)))
        cluster (Cluster/get system)
        client (-> system
                   (ClusterClientSettings/create)
                   (.withInitialContacts
                     #{(ActorPaths/fromString (format "akka.tcp://market@192.168.50.83:%d/system/receptionist" port))})
                   (ClusterClient/props)
                   (#(.actorOf system % "client")))
        service-actor (.actorOf system (ClojureActor/propsWithInit
                                         initer
                                         service) "service")]
    (try
      (doto (ClusterClientReceptionist/get system)
        (.registerSubscriber channel self)
        (.registerService service-actor))
      (Thread/sleep 100)
      (.tell client (ClusterClient$Publish. channel empty-depth) self)
      (await)
      (.expectMsgPF test-kit "expect empty depth from client publish"
                    (function [msg]
                      (is (and (instance? Depth msg)
                               (= empty-depth msg)))))
      (.tell client (ClusterClient$Send. "/user/service" empty-depth) self)
      (await)
      (.expectMsgPF test-kit "expect empty depth from client send"
                    (function [msg]
                      (is (and (instance? Depth msg)
                               (= empty-depth msg)))))
      (finally
        (do
          (doto (ClusterClientReceptionist/get system)
            (.unregisterService service-actor)
            (.unregisterSubscriber channel self))
          (.leave cluster (.selfAddress cluster))
          (TestKit/shutdownActorSystem system))))))


(deftest cross-feedback-test
  "startup a inner actor and link to cluster, publish message and get feedback"
  (let [channel "btcusdt.depth.step0"
        system (ActorSystem/create "market")
        test-kit (TestKit. system)
        self (.getRef test-kit)
        port (-> system
                 (.provider)
                 (.getDefaultAddress)
                 (.port)
                 (.get))
        empty-depth (doto (Depth.)
                      (.setChannel "btcusdt.depth.step0")
                      (.setVersion 0)
                      (.setAsk [])
                      (.setBid []))
        await #(.awaitCond test-kit (supplier [_] (.msgAvailable test-kit)))
        cluster (Cluster/get system)
        client (-> system
                   (ClusterClientSettings/create)
                   (.withInitialContacts
                     #{(ActorPaths/fromString (str "akka.tcp://market@192.168.50.83:" port "/system/receptionist"))})
                   (ClusterClient/props)
                   (#(.actorOf system % "client")))
        service-actor (.actorOf system (ClojureActor/propsWithInit
                                         initer
                                         service) "service")
        mediator (-> system
                     (DistributedPubSub/get)
                     (.mediator))]
    (try
      (-> (DistributedPubSubMediator$Subscribe. channel self)
          (#(.tell mediator % self)))
      (await)
      (.expectMsgPF test-kit "subscribe ack"
                    (function [msg]
                      (is (instance? DistributedPubSubMediator$SubscribeAck msg))))
      (doto (ClusterClientReceptionist/get system)
        (.registerService service-actor))
      (.tell client (ClusterClient$Publish. channel empty-depth) self)
      (await)
      (.expectMsgPF test-kit "expect empty depth from client publish"
                    (function [msg]
                      (is (and (instance? Depth msg)
                               (= empty-depth msg)))))
      (.tell client (ClusterClient$Send. "/user/service" empty-depth) self)
      (await)
      (.expectMsgPF test-kit "expect empty depth from client send"
                    (function [msg]
                      (is (and (instance? Depth msg)
                               (= empty-depth msg)))))
      (-> (DistributedPubSubMediator$Unsubscribe. channel self)
          (#(.tell mediator % self)))
      (await)
      (finally
        (do
          (doto (ClusterClientReceptionist/get system)
            (.unregisterService service-actor)
            (.unregisterSubscriber channel self))
          (.leave cluster (.selfAddress cluster))
          (TestKit/shutdownActorSystem system))))))

File: robot-seed/src/test/java/liu/mars/market/ListenerTest.java
package liu.mars.market;

import akka.actor.*;
import akka.cluster.Cluster;
import akka.cluster.client.ClusterClient;
import akka.cluster.client.ClusterClientReceptionist;
import akka.cluster.client.ClusterClientSettings;
import akka.cluster.pubsub.DistributedPubSub;
import akka.cluster.pubsub.DistributedPubSubMediator;
import akka.testkit.javadsl.TestKit;
import liu.mars.market.dash.Depth;
import org.junit.AfterClass;
import org.junit.Assert;
import org.junit.BeforeClass;
import org.junit.Test;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.Set;

public class ListenerTest {
    public static class Listener extends AbstractActor {
        private String feedback;
        private Listener(String feedback){
            this.feedback = feedback;
        }

        @Override
        public Receive createReceive() {
            return receiveBuilder().match(Depth.class, msg -> {
                this.context().system().actorSelection(String.format("/system/%s", feedback))
                        .tell(msg, self());
            }).build();
        }


        public static Props props(String feedback) {
            return Props.create(Listener.class, ()-> new Listener(feedback));
        }
    }

    private static ActorSystem system;
    private static Cluster cluster;
    private static String channel = "btcusdt.depth.step0";


    static private Set<ActorPath> initialContacts(Integer port){
        Set<ActorPath> re = new HashSet<>();
        re.add(ActorPaths.fromString(String.format("akka.tcp://market@192.168.50.83:%d/system/receptionist", port)));
        return re;
    }

    private Depth emptyDepth() {
        var re = new Depth();
        re.setAsk(new ArrayList<>());
        re.setBid(new ArrayList<>());
        re.setChannel(channel);
        re.setVersion(0);
        re.setTs(LocalDateTime.now());
        return re;
    }

    @BeforeClass
    public static void setup() throws InterruptedException {
        system = ActorSystem.create("market");
        cluster = Cluster.get(system);
    }

    @AfterClass
    public static void teardown() throws InterruptedException {
        cluster.leave(cluster.selfAddress());
        TestKit.shutdownActorSystem(system);
        system = null;
    }

    @Test
    public void testPubSub() {
        new TestKit(system){{
            var mediator = DistributedPubSub.get(system).mediator();
            var self = getRef();

            mediator.tell(new DistributedPubSubMediator.Subscribe(channel, self), self);
            this.awaitCond(this::msgAvailable);
            this.expectMsgPF("expect subscribed ack", msg -> {
                Assert.assertTrue(msg instanceof DistributedPubSubMediator.SubscribeAck);
                return msg;
            });

            mediator.tell(new DistributedPubSubMediator.Publish(channel, emptyDepth()), self);
            this.awaitCond(this::msgAvailable);
            this.expectMsgPF("expect empty depth", msg -> {
                Assert.assertTrue(msg instanceof Depth);
                var depth = (Depth)msg;
                Assert.assertEquals(0, depth.getVersion());
                Assert.assertEquals(0, depth.getAsk().size());
                Assert.assertEquals(0, depth.getBid().size());
                Assert.assertEquals(channel, depth.getChannel());

                return msg;
            });


            mediator.tell(new DistributedPubSubMediator.Unsubscribe(channel, self), self);
            this.awaitCond(this::msgAvailable);
            this.expectMsgPF("expect unsubscribed ack", msg -> {
                Assert.assertTrue(msg instanceof DistributedPubSubMediator.UnsubscribeAck);
                return msg;
            });

        }};
    }

    @Test
    public void testClient() throws InterruptedException {
        new TestKit(system) {{
            var self = getRef();
            var receptionlist = ClusterClientReceptionist.get(system);
            var serviceActor = system.actorOf(Listener.props(self.path().name()), "service-client");
            Integer port = (Integer) system.provider().getDefaultAddress().port().get();
            var client = system.actorOf(
                    ClusterClient.props(ClusterClientSettings.create(system)
                            .withInitialContacts(initialContacts(port))));
            try {
                receptionlist.registerService(serviceActor);
                receptionlist.registerSubscriber(channel, self);
                Thread.sleep(1000);
                client.tell(new ClusterClient.Publish(channel, emptyDepth()), self);
                this.awaitCond(this::msgAvailable);
                expectMsgPF("expect empty depth from client publish", msg -> {
                    Assert.assertTrue(msg instanceof Depth);
                    var depth = (Depth) msg;
                    Assert.assertEquals(0, depth.getVersion());
                    Assert.assertEquals(0, depth.getAsk().size());
                    Assert.assertEquals(0, depth.getBid().size());
                    Assert.assertEquals(channel, depth.getChannel());
                    return msg;
                });

                client.tell(new ClusterClient.Send("/user/service-client", emptyDepth()), self);
                this.awaitCond(this::msgAvailable);
                expectMsgPF("expect empty depth from client publish", msg -> {
                    Assert.assertTrue(msg instanceof Depth);
                    var depth = (Depth) msg;
                    Assert.assertEquals(0, depth.getVersion());
                    Assert.assertEquals(0, depth.getAsk().size());
                    Assert.assertEquals(0, depth.getBid().size());
                    Assert.assertEquals(channel, depth.getChannel());
                    return msg;
                });
            }finally {
                receptionlist.unregisterService(serviceActor);
                receptionlist.unregisterSubscriber(channel, self);
            }
        }};
    }

    @Test
    public void testCross() {
        new TestKit(system) {{
            var self = getRef();

            var mediator = DistributedPubSub.get(system).mediator();
            var serviceActor = system.actorOf(Listener.props(self.path().name()), "service-cross");
            Integer port = (Integer) system.provider().getDefaultAddress().port().get();
            var client = system.actorOf(
                    ClusterClient.props(ClusterClientSettings.create(system)
                            .withInitialContacts(initialContacts(port))));

            mediator.tell(new DistributedPubSubMediator.Subscribe(channel, serviceActor), self);
            awaitCond(this::msgAvailable);
            expectMsgPF("expect subscribed ack", msg -> {
                Assert.assertTrue(msg instanceof DistributedPubSubMediator.SubscribeAck);
                return msg;
            });

            ClusterClientReceptionist.get(system).registerService(serviceActor);
            client.tell(new ClusterClient.Publish(channel, emptyDepth()), self);
            awaitCond(this::msgAvailable);
            expectMsgPF("expect empty depth from client publish", msg -> {
                Assert.assertTrue(msg instanceof Depth);
                var depth = (Depth)msg;
                Assert.assertEquals(0, depth.getVersion());
                Assert.assertEquals(0, depth.getAsk().size());
                Assert.assertEquals(0, depth.getBid().size());
                Assert.assertEquals(channel, depth.getChannel());
                return msg;
            });

            mediator.tell(new DistributedPubSubMediator.Publish(channel, emptyDepth()), self);
            this.awaitCond(this::msgAvailable);
            expectMsgPF("expect empty depth from client publish", msg -> {
                Assert.assertTrue(msg instanceof Depth);
                var depth = (Depth)msg;
                Assert.assertEquals(0, depth.getVersion());
                Assert.assertEquals(0, depth.getAsk().size());
                Assert.assertEquals(0, depth.getBid().size());
                Assert.assertEquals(channel, depth.getChannel());
                return msg;
            });


            mediator.tell(new DistributedPubSubMediator.Unsubscribe(channel, serviceActor), self);
            awaitCond(this::msgAvailable);
            expectMsgPF("expect unsubscribed ack", msg -> {
                Assert.assertTrue(msg instanceof DistributedPubSubMediator.UnsubscribeAck);
                return msg;
            });
        }};
    }

    @Test
    public void testRouter() throws InterruptedException {
        new TestKit(system) {{
            var self = getRef();
            var receptionlist = ClusterClientReceptionist.get(system);
            Integer port = 2558;
            var client = system.actorOf(
                    ClusterClient.props(ClusterClientSettings.create(system)
                            .withInitialContacts(initialContacts(port))));
            try {
                receptionlist.registerSubscriber(channel, self);
                Thread.sleep(1000);
                client.tell(new ClusterClient.Publish(channel, emptyDepth()), self);
                this.awaitCond(this::msgAvailable);
                expectMsgPF("expect empty depth from client publish", msg -> {
                    Assert.assertTrue(msg instanceof Depth);
                    var depth = (Depth) msg;
                    Assert.assertEquals(0, depth.getVersion());
                    Assert.assertEquals(0, depth.getAsk().size());
                    Assert.assertEquals(0, depth.getBid().size());
                    Assert.assertEquals(channel, depth.getChannel());
                    return msg;
                });

                client.tell(new ClusterClient.Send("/user/seed", emptyDepth()), self);
                this.awaitCond(this::msgAvailable);
                expectMsgPF("expect empty depth from client publish", msg -> {
                    Assert.assertTrue(msg instanceof Depth);
                    var depth = (Depth) msg;
                    Assert.assertEquals(0, depth.getVersion());
                    Assert.assertEquals(0, depth.getAsk().size());
                    Assert.assertEquals(0, depth.getBid().size());
                    Assert.assertEquals(channel, depth.getChannel());
                    return msg;
                });
            }finally {
                receptionlist.unregisterSubscriber(channel, self);
            }
        }};
    }
}
